keps/0027-force-pod-reload-parameters.md

```diff
@@ -0,0 +1,138 @@
+---
+kep-number: 27
+short-desc: A parameter flag to disable forced pod reload
+title: Forced Pod reload attribute
+authors:
+  - @aneumann82
+owners:
+  - @aneumann82
+editor:
+creation-date: 2020-03-31
+last-updated: 2020-03-31
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+## Summary
+
+This KEP describes addition of a flag for parameters that controls the forced pod reloading
+
+## Motivation
+
+The new flag will allow an operator developer to control that a specific attribute does not automatically restart
+all pods of a deployment or stateful set if this parameter is changed.
+
+### Goals
+
+Make it possible for an operator developer to specify that a parameter will *not* automatically restart all pods from
+a stateful set or a deployment.
+
+### Non-Goals
+
+- Detailed control when Pods are restarted: If the spec template for a stateful set is changed, Kubernetes will still
+reload all pods automatically.
+- Calculation of used config maps, secrets or other resources used by a pod: The automatic pod restart is currently 
+used to force pods to re-read config maps on a parameter update. When a user explicitly disables the pod restart for 
+a parameter, and this parameter updates a config map or another dependency of the pod, the change might not be applied
+in the pod until the next restart.
+
+## Proposal
+
+Add an additional attribute `forcePodRestart` to parameter specifications in `params.yaml`:
+
+```yaml
+  - name: NODE_COUNT
+    description: "Number of Cassandra nodes."
+    default: "3"
+    forcePodRestart: "false"
+```
+
+The default value for this parameter would be `true` to keep backwards compatibility. The general behavior should stay
```

Comment:
I'm all for keeping backwards compatibility. On the other hand I feel that enabling forced pod restarts is a feature that users have to opt-in, i.e. the default should be `false`. Forced pod restarts enable updating config maps on parameter updates with the caveat of changing some assumptions regarding pod lifetimes. As such the default should be what would be the expected behavior of pods not restarting.

Reply:
I'm not sure. I think the pod restarting is something that you "usually" want. There may be applications that are "ConfigMap-Aware" and re-read config files without a pod restart. For these, a default of "false" would be ok, but I think most applications will require the pod restarts for most of the parameters.

Having this as "false" for default will probably lead to weird issues for operator developers where they wonder why they've updated a parameter, but the running application keeps running with an old value.

This sounds like something that would be useful to set as a "default" for all parameters, but I'm not sure adding that complexity in would be good.

Reply:
@zmalik what would you say should be the default?

Reply:
@ANeumann82  before we decide how to change the current behavior, I would like to understand better why it's there in the first place.

Do we have a record of the initial motivation/discussion/KEP for this `kudo.dev/last-plan-execution-uid` feature? It seems like you're implying that this was an intentional mechanism put in place to force pod restarts on any parameter change or upgrade. Did I read that right?

Or perhaps you are just describing the current effect, but this was not a designed feature but just an accident - i.e. we were trying to mimic what kustomize was doing?

Given the above lack of information, it is hard to judge but I have to say that in my opinion this feature is surprising in the negative sense:
* it is not documented anywhere, as far as I can tell,
* it changes the default behaviour of k8s statefulsets and deployments,
* the name of the annotation does not provide any hint about for its (implied? intent) - for example when I first saw this annotation being changed very recently (when debugging the issue described in https://github.com/mesosphere/kudo-cassandra-operator/pull/78) - it was far from obvious to me why this is happening
* there is a serious scoping issue - it affects all statefulsets/deployments in an operator regardless of whether they actually use the parameter at all. Yes, there are ways to work that around by triggering a special plan, but this also has issues.

Adding a `forcePodReload` flag builds more complexity on an unstable foundation, which I think is exactly what we should be avoiding while we're still in `v0...` stage.

In general I'd vote for breaking compatibility in this case and get rid of this feature altogether. We did break compatibility for more minor reasons in the past.

If an operator developer wants a certain statefulset/deployment to restart on a parameter change this can be achieved in another way, either using the tool you mentioned, or by us providing a hash function that can be used as seen below by an operator developer to *explicitly* declare which parameters require a restart:

```
spec:
  template:
    metadata:
      annotations:
        config-values-hash: {{ hash(
          .Params.SOMETHING_THAT_AFFECTS_THIS_POD,
          .Params.SOMETHING_ELSE_THAT_AFFECTS_THIS_POD,
           ...
          ) }}
```

Reply:
To give this discussion some context: the feature (though not documented) is very much intentional. Originally, we used [stakater/Reloader](https://github.com/stakater/Reloader) but it came with its [own issues](https://github.com/kudobuilder/kudo/issues/1025) so we decided to implement our own solution. Which right now is applied to [all templates](https://github.com/kudobuilder/kudo/issues/1036) and surely can be smarter by updating/restarting only things that are affected.

Reply:
Thanks for the links! Although I fail to see if https://github.com/kudobuilder/kudo/issues/1025 was investigated to find the root cause? Maybe it's there in the description, but only obvious for the initiated :-)
Also I'm surprised to see so little discussion on https://github.com/kudobuilder/kudo/pull/1031 :thinking:

Reply:
Maybe @alenkacz or @gerred could shed additional light on it 😉

Reply:
@porridge Thanks for that comment! I think @zen-dog answered some of the questions about the origin of this issue.

I kind of like your approach with the hash, although I don't like to put it directly into the statefulset-configuration: If we think about an operator with > 150 Parameters, the call to `hash( PARAM1, PARAM2, ..., PARAMN) would get really long, and require a operator developer to touch at least two or more places when adding a parameter.

I've added a second proposal to the document that takes up the idea of a hash calculation based on parameters, but have the hash-params defined on the parameters itself, please have a look.

Reply:
Well, the whole point of a hash is that it can be of arbitrarily short length, not dependent on the number of arguments :-)